core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -344,15 +338,20 @@ private[spark] class BlockStatusesAccumulator
   override def add(v: (BlockId, BlockStatus)): Unit = _seq.add(v)
 
   override def merge(
-    other: AccumulatorV2[(BlockId, BlockStatus), java.util.List[(BlockId, BlockStatus)]]): Unit = {
+    other: AccumulatorV2[(BlockId, BlockStatus), List[(BlockId, BlockStatus)]]): Unit = {
     other match {
-      case o: BlockStatusesAccumulator => _seq.addAll(o.value)
+      case o: BlockStatusesAccumulator => o.value.foreach(_seq.add)
       case _ => throw new UnsupportedOperationException(
         s"Cannot merge ${this.getClass.getName} with ${other.getClass.getName}")
     }
   }
 
-  override def value: java.util.List[(BlockId, BlockStatus)] = _seq
+  // `asScala` accesses the internal values using `java.util.Iterator` so needs to be synchronized
+  override def value: List[(BlockId, BlockStatus)] = {
+    _seq.synchronized {
```

Comment:
Thanks for the review @zsxwing . I checked the java doc and it says that getting iterator is not thread safe and suggests below usage. That's why I did `_seq.synchronized`

https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html

```
  List list = Collections.synchronizedList(new ArrayList());
      ...
  synchronized (list) {
      Iterator i = list.iterator(); // Must be in synchronized block
      while (i.hasNext())
          foo(i.next());
  }
 
```

Regarding your second questions, `JsonProtocal` is using it as Scala collection that is why I converted it to a Scala collection so we won't need to convert again.